src/herder/TxSetUtils.cpp

```diff
@@ -0,0 +1,389 @@
+                if (returnEarlyOnFirstInvalidTx)
+                {
+                    if (minSeqCheckIsInvalid)
+                    {
+                        CLOG_DEBUG(Herder,
+                                   "minSeqAge or minSeqLedgerGap set on tx "
+                                   "without lowest seqNum. tx: {}",
+                                   xdr_to_string(tx->getEnvelope(),
+                                                 "TransactionEnvelope"));
+                    }
+                    else
+                    {
+                        CLOG_DEBUG(
+                            Herder,
+                            "Got bad txSet: {} tx invalid lastSeq:{} tx: {} "
+                            "result: {}",
+                            hexAbbrev(txSet.previousLedgerHash()), lastSeq,
+                            xdr_to_string(tx->getEnvelope(),
+                                          "TransactionEnvelope"),
+                            tx->getResultCode());
+                    }
+                    return invalidTxs;
+                }
+            }
+            else // update the account fee map
+            {
+                lastSeq = tx->getSeqNum();
+                int64_t& accFee = accountFeeMap[tx->getFeeSourceID()];
+                if (INT64_MAX - accFee < tx->getFeeBid())
+                {
+                    accFee = INT64_MAX;
+                }
+                else
+                {
+                    accFee += tx->getFeeBid();
+                }
+                ++iter;
+            }
+        }
+    }
+
+    auto header = ltx.loadHeader();
+    for (auto& kv : accountTxMap)
+    {
+        auto iter = kv.second.begin();
+        while (iter != kv.second.end())
+        {
+            auto tx = *iter;
+            auto feeSource = stellar::loadAccount(ltx, tx->getFeeSourceID());
+            auto totFee = accountFeeMap[tx->getFeeSourceID()];
+            if (getAvailableBalance(header, feeSource) < totFee)
+            {
+                while (iter != kv.second.end())
+                {
+                    invalidTxs.emplace_back(*iter);
+                    ++iter;
+                }
+                if (returnEarlyOnFirstInvalidTx)
+                {
+                    CLOG_DEBUG(Herder,
+                               "Got bad txSet: {} account can't pay fee tx: {}",
+                               hexAbbrev(txSet.previousLedgerHash()),
+                               xdr_to_string(tx->getEnvelope(),
+                                             "TransactionEnvelope"));
+                    return invalidTxs;
+                }
+            }
+            else
+            {
+                ++iter;
+            }
+        }
+    }
+
+    return invalidTxs;
+}
+
+// Target use case is to remove a subset of invalid transactions from a TxSet.
+// I.e. txSet.size() >= txsToRemove.size()
+TxSetFrameConstPtr
+removeTxs(TxSetFrameConstPtr txSet, TxSetFrame::Transactions const& txsToRemove)
```

Comment:
mark as "static" or in the `{}` namespace as to not export

Reply:
actually, I see that you did something odd wrt test code. I am not sure if there is a way to do this with namespaces, but with classes this would done by having that defined as a private static method and have the test class defined as a "friend"

Reply:
we want to make it basically impossible to call this function from production code outside of this file, and what you did does not guarantee this and is definitely not a usual pattern